### ISSUE TITLE
How to resume training from already trained spark ner model.

### ISSUE THREAD

**Issue by Harish202** [JohnSnowLabs/spark-nlp#1028]
First of all thank you for such a amazing repo. You guys are amazing. I have few questions and hopefully you can guide me. 

How do I resume training from a given trained spark ner model. And also how does training limit work. Say if limit is 1000 will it randomly sample 1000 from training data or will it take first 1000 training data samples. I am using Python version of sparknlp bert model.

And I tried going through documentation. It would be great if I can get a link where I can get to know all the available arguments for the .fit method for training in python with corresponding description. Unfortunately I am not able to find those details. Please let me know if there is any link with such detailed documentation. Appreciate your help.

**Comment by maziyarpanahi** [comment]
Hi @Harish202 

> How do I resume training from a given trained spark ner model. And also how does training limit work. Say if limit is 1000 will it randomly sample 1000 from training data or will it take first 1000 training data samples. I am using Python version of sparknlp bert model.

We don't have a resume on our pretrained models. The weights are constant once the model is saved. In theory, it is very possible for us to implement this resume as we load the weights, but we haven't found any practical evidence that someone can resume something over, for instance, CoNLL2003 models in NER and either learn new things or improve it. That's being said, the resume only works with NerDL, ClassifierDL, MultiClassifierDL, SentimentDL, etc. The Transformers are pretrained weights that cannot be resumed. (we are working on having a new annotator for fine-tuning BERT for some tasks, but that's for BERT which comes with lots of parameters and information.)

Since we have a large number of Embeddings for both Word and Sentence, we use state-of-the-art deep learning algorithms in our models to take advantage of transfer learning from these embeddings. In most situations, you may not even need a large training example to have a good model thanks to those embeddings and the DL architecture behind the model.

We don't have a limit in any of the training, whatever you provide will be part of the training examples, if you wish to choose a fraction of it for validation during the training for self-evacuation it will shuffle and randomly takes less say 0.1 of that training examples.

> And I tried going through documentation. It would be great if I can get a link where I can get to know all the available arguments for the .fit method for training in python with corresponding description. Unfortunately I am not able to find those details. Please let me know if there is any link with such detailed documentation. Appreciate your help.

Each annotator that can be trained has its own set of parameters. In Apache Spark, the .fit() and .transform() only accepts one argument which is the DataFrame. All the hyperparameters are set during the initialization of the annotator. For instance, for training NER: 

```scala
val ner = new NerDLApproach()
.setInputCols("sentence", "token", "embeddings")
.setOutputCol("ner")
.setLabelColumn("label")
.setOutputCol("ner")
.setLr(1e-3f) //0.001
.setPo(5e-3f) //0.005
.setDropout(5e-1f) //0.5
.setBatchSize(128)
.setMaxEpochs(50)
.setRandomSeed(0)
.setVerbose(0)
.setEvaluationLogExtended(true)
```
We have a very nice course for Apache Spark/Spark NLP in Python here: (all notebooks can be executed right in the Google Colab)
https://github.com/JohnSnowLabs/spark-nlp-workshop/tree/master/tutorials/Certification_Trainings/Public

**Comment by Harish202** [comment]
Thank you for the detailed explanation. Actually my problem is due to limited ram space in colab my sparknlp ner training is getting killed. So what I thought is instead of giving all of the training data at once,  if trainingdata.limit(1000) will usually sample 1000 data randomly, then perhaps I can train with limit in place. So that over the course of total epochs I can hope that all training data would be covered. I already tried to reduce batch size and all but to no effect. Can you please let me know if trainingdata.limit(1000) make a random sample per epoch or it just take first 1000 for all epochs and will continue for training?

**Comment by maziyarpanahi** [comment]
Yes, that is a good point. The limit doesn't do any randomization or shuffling, it will just cut the data at a certain row number. I would suggest the following for randomization:

```python
randomizedTrainingDF = myFullTrainingDF.orderBy(rand())
```
(rand is in Spark ML)

Then the following for a better split:

```python
trainDataset, smallerTrainDataset = randomizedTrainingDF.randomSplit([0.8, 0.2], seed = 12345)
```
One side note, after the split, you need to make sure your labels/classes are fairly distributed. A simple explode/groupBy should give you that info to be sure you have all the labels available and they all have enough examples.

As a final note, you may want to consider some of the newest BERT Embeddings introduced in 2.6.0 such as `small_bert_L2_128` which is 15MB and over 20x faster. 

https://github.com/JohnSnowLabs/spark-nlp-models#embeddings

**Comment by Harish202** [comment]
Wow, that helps a lot buddy. Thank you for taking time. Very much appreciated. God bless :) . I will close the issue. Good day 👍

**Comment by joao-prg** [comment]
Hi @maziyarpanahi !
I would like to clarify this 

> We don't have a resume on our pretrained models. The weights are constant once the model is saved. In theory, it is very possible for us to implement this resume as we load the weights, but we haven't found any practical evidence that someone can resume something over, for instance, CoNLL2003 models in NER and either learn new things or improve it. That's being said, the resume only works with NerDL, ClassifierDL, MultiClassifierDL, SentimentDL, etc. 

I'm trying to train a multi-class text classifier, save it, and then load it to perform fine tuning. Is this possible?

Thanks in advance!

**Comment by maziyarpanahi** [comment]
Hi @joao-prg 
We don't have any resuming nor fine-tuning of already trained models of all kinds. Once any model is trained until the last Epoch, whether it is saved or in memory, it cannot resume or used for any purpose rather than prediction/inference.

**Comment by joao-prg** [comment]
Thanks for the clarification @maziyarpanahi !
